### ISSUE TITLE
Manage Document fields

### ISSUE THREAD

**Issue by quartzmo** [GoogleCloudPlatform/gcloud-ruby#413]
Document fields are probably the most complex part of the Search v1 REST API. We need to decide the best way for the gcloud-ruby API to manage fields and their values.

## Background

The following excerpts were **both** taken from the Cloud Search guide to [Documents and Indexes](https://cloud.google.com/search/documents_indexes):

In the first mention of fields, `name`, `type`, and `value` are attributes of each field.

> A document is an object with a unique ID and a list of fields containing user data. Each field has a name and a type. There are several types of fields, identified by the kinds of values they contain:
> 
> Atom Field - an indivisible character string
> Text Field - a plain text string that can be searched word by word
> HTML Field - a string that contains HTML markup tags, only the text outside the markup tags can be searched
> Number Field - a floating point number
> Date Field - a date object with year/month/day and optional time
> Geopoint Field - a data object with latitude and longitude coordinates  

In the second mention of fields, `name` is a unique attribute of each field, which may contain multiple values of different types.

> A document contains a list of one or more fields, with each field having a name and a list of values. Each field name consists of only ASCII characters, must be unique within the document and is case sensitive. A field name must start with a letter and can contain letters, digits, or underscore, with a maximum of 500 characters. A field can have multiple values with same or different types; however, it cannot have multiple Timestamp or number values.

Obviously, the data in both cases can be the same. It is just a question of how best to manage it.

#### Option 1: Grouped by name

In option 1, multiple values are set on a Field object that is created (or retrieved?) by name. This is the
approach taken by the REST API as well as gcloud-python.

```ruby
document = index.document "product-sku-000001"

field = document.field "description"
field.add_value "100% organic cotton ruby gem T-shirt",
                type: :text,
                lang: "en"
field.add_value "<p>100% organic cotton ruby gem T-shirt</p>",
                type: :html,
                lang: "en"
field.values.count #=> 2
```

#### Option 2: Flat

In option 2, multiple fields are added to a single list. Field `name` is not unique.

```ruby
document = index.document "product-sku-000001"

document.add_field "description",
                   "100% organic cotton ruby gem T-shirt",
                   type: :text,
                   lang: "en"

document.add_field "description",
                   "<p>100% organic cotton ruby gem T-shirt</p>",
                   type: :html,
                   lang: "en"
document.fields_by(name: "description").count #=> 2
```

An advantage of the flat structure is that you don't have to go through the extra step of creating/retrieving a collecting Field object in order to add a field value. A disadvantage is that you have to provide the name to each `add_field` call. Another disadvantage is that it is a little harder to inspect the values you have already added.

#### Option 1: Grouped by name

Fields are returned as a hash of objects, with `name` values as keys. This is the approach taken by the REST API as well as gcloud-python.

```ruby
document = index.document "product-sku-000001"

field = document.fields["description"]

field.values[0].field.name #=> "description"
field.values[0].value #=> "100% organic cotton ruby gem T-shirt"
field.values[0].type #=> :text
field.values[0].lang #=> "en"
field.values[1].field.name #=> "description"
field.values[1].value #=> "<p>100% organic cotton ruby gem T-shirt</p>"
field.values[1].type #=> :html
field.values[1].lang #=> "en"
```

In practice, I have found that pre-grouped data structures often do more harm than good. An example is the [Zonefile](https://github.com/boesemar/zonefile) library, which returns RRs in a hash grouped by type. The RR values in the hash do *not* contain the type, so converting to a flat structure that includes both the type and the other attributes in each element is a chore.

Thus, in this example, note that each value contains a reference back to its field. (Without it or at least the copied `name` value, building other data structures using the values alone is problematic, since the names are lost.)

#### Option 2: Flat

Fields are returned flat, in a single array, with non-unique name. They can be filtered (grouped) in Ruby using `Document#fields_by` or `Array#select`.

```ruby
index = search.index "products"
query = search.query "t-shirt"
document = index.document "product-sku-000001"

fields = document.fields_by(name: "description")
# Same as
fields = document.fields.select {|f| f.name == "description"}

fields[0].name #=> "description"
fields[0].value #=> "100% organic cotton ruby gem T-shirt"
fields[0].type #=> :text
fields[0].lang #=> "en"
fields[1].name #=> "description"
fields[1].value #=> "<p>100% organic cotton ruby gem T-shirt</p>"
fields[1].type #=> :html
fields[1].lang #=> "en"
```

An advantage of the flat structure is that the fields can just as easily be filtered by a different attribute. In fact, given Ruby's rich capabilities for programming in a functional style, you easily do quite a lot given a flat list of values. For example, if we want a list of the document's English HTML fields:

```ruby
index = search.index "products"
query = search.query "t-shirt"
document = index.document "product-sku-000001"

fields = document.fields_by(type: :html, lang: "en")
# Same as
fields = document.fields.select {|f| f.type == :html && f.lang == "en"}

fields[0].name #=> "description"
fields[0].value #=> "<p>100% organic cotton ruby gem T-shirt</p>"
fields[0].type #=> :html
fields[0].lang #=> "en"
```

**Comment by blowmage** [comment]
One thing I do not like about some other language code examples is explicitly creating a `field` object. FWIW, I would like the document to behave as close to a hash with a default value of an empty array:

```ruby
Hash.new { |hash,key| hash[key] = [] }
```

This allows the creation of fields (keys) on the fly, as well as supports multiple values for the field.

```ruby
document = Hash.new { |hash,key| hash[key] = [] }
document["title"] << "Hello Gcloud!"
document["body"] << "gcloud is a client library for accessing Google Cloud Platform services that ..."
document["body"] << "<p><code>gcloud</code> is a client library for accessing Google Cloud Platform services that ...</p>"
```

Retrieving the field values would also be very similar.

```ruby
document = Hash.new { |hash,key| hash[key] = [] }
title = document["title"].first if document["title"].any?
body = document["body"].first if document["body"].any?
```

Notice we aren't calling `.value` to get the field value. I think we should wrap the raw value in some `FieldValue` class that delegates to the value, but adds the `lang` and `type` attributes. 

```ruby
body = document["body"].first if document["body"].any?
body.lang #=> :en
body.type #=> :text
```

And then we add all the helpers that make it easy to retrieve the field values by `lang`/`type` combination, like the ones that you have already mentioned. We also might consider adding some helpers that make it explicit what type of value we are adding:

 ```ruby
document = Gcloud::Search::Document.new
document["body"].add_html "<p><code>gcloud</code> estas kliento biblioteko por aliranta Google Nubo Platformo servoj kiu ...</p>"
```

**Comment by blowmage** [comment]
The other thought I've been having about the search documents is the JSON API document structure:

```json
{
  "docId": "doc123",
  "rank": 123456789,
  "fields": {
    "title": {
      "values": [
        {
          "stringFormat": "TEXT",
          "lang": "en",
          "stringValue": "Hello Gcloud!"
        }
      ]
    },
    "body": {
      "values": [
        {
          "stringFormat": "TEXT",
          "lang": "en",
          "stringValue": "gcloud is a client library for accessing Google Cloud Platform services that ..."
        },
        {
          "stringFormat": "HTML",
          "lang": "en",
          "stringValue": "<p><code>gcloud</code> is a client library for accessing Google Cloud Platform services that ...</p>"
        },
        {
          "stringFormat": "HTML",
          "lang": "eo",
          "stringValue": "<p><code>gcloud</code> estas kliento biblioteko por aliranta Google Nubo Platformo servoj kiu ...</p>"
        }
      ]
    }
  }
}
```

I think that we should allow both importing and exporting JSON strings and Hash objects that adhere to this structure. Working with a document object would be easier for sure, but I think there are advantages to allowing folks to use this data structure as well. If they want to.

```ruby
document = index.document "doc123"
document.to_json #=> "{\"docId\": \"doc123\",\"rank\": 123456789, ... }"
document.to_hash #=> {"docId"=>"doc123","rank"=>123456789, ... }

index.import_json "doc456", "{\"docId\": \"doc456\",\"rank\": 123456790, ... }"
index.import_hash "doc789", {"docId"=>"doc789","rank"=>123456791, ... }
```

**Comment by quartzmo** [comment]
These are great ideas. 

I really agree with leveraging Hash and Array wherever possible. My only "must have" for accessing Document fields, which fits in just fine with the Hash/Array structure, is that the `name` attribute also be available on the `FieldValue` objects. Without it, you can't create new collections using the values of multiple fields, because you lose knowledge of to which fields the values belong.

Here's a scenario. Imagine you want the English-language `html` values from all fields that have them, to pass to a template for rendering:

```ruby
html_values = document.values.flatten.select { |v| v.type == :html && v.lang == :en } 
html_values.count #=> 3
html_values[0].name #=> "title"
html_values[1].name #=> "summary"
html_values[2].name #=> "body"
```

If the `name` is not available on the values, you can't do cool functional-style transformations like this.

**Comment by quartzmo** [comment]
@blowmage Do you want to extract your import/export description to a new issue? Seems to merit it.

**Comment by blowmage** [comment]
Sure.

**Comment by quartzmo** [comment]
@blowmage I'm going to start work on this issue as a `Fields` module, that can be mixed into both `Document` and `Result`. If the module doesn't seem worth it by the time `Document` is merged, I will refactor.

I'm going to follow your suggestions for hash key accessors and values that delegate to their simple values while implementing `lang` and `type`.

**Comment by blowmage** [comment]
Here is what I'm thinking for adding field values with the optional `type` and `lang` values. We should start with an `#add` method that takes the name, value, and optional arguments for type and lang:

```ruby
doc.add :title, "<h1>Hello world!</h1>", lang: :en, type: :html
```

Ince we have that I think we should also add `#type` and `#lang` filter methods. This would filter fields by those values, but also allow new values to be added:

```ruby
# get all the fields that are html and english
doc.type(:html).lang(:en)

# set a new value for html in english for the field title
doc.type(:html).lang(:en)["title"] = "<h1>Hello world!</h1>"
```

I suspect this API would be fun to use, and offer new flexible ways to deal with and manage the data. Hopefully this makes the simple things simple, but the more complex things remain possible.

**Comment by quartzmo** [comment]
Sounds good. I'll start with a commit for `Document#add` after I rebase PR #433 on PR #427, and at the same time push `Fields#[]=` (field setter without `type` or `lang` options).

**Comment by blowmage** [comment]
Here is my list of the things that are still outstanding for fields and field values:

- [ ] Remove a field from a document
- [ ] Add multiple values to a field on a document
- [ ] Specify the language when adding a field value on a document
- [ ] Specify the type when adding a field value on a document (text, html, atom, den, etc...)
- [ ] Remove a value from the list of values on a document's field
- [ ] Iterate through all the field values on a Document/Result
- [ ] Identify the language of a field value on a Document/Result
- [ ] Identify the type of a field value on a Document/Result

**Comment by blowmage** [comment]
Closed by #440.
